Developer Journey Level 5 (#2210)
dfinity/portal: docs/tutorials/developer-journey/level-5/5.5-SNS-tutorial.md

# 5.5 Launching an SNS

## Overview

A Service Nervous System (SNS) is an advanced form of a DAO. If you recall, DAOs and the NNS were covered in a previous module, [4.4 NNS and governance](../level-4/4.4-nns-governance.md). To recap, a DAO, or decentralized autonomous organization, is a form of decentralized governance that is controlled by the DAO's stakeholders without a central form of authority. The Internet Computer is managed by a DAO known as the Network Nervous System. 

> The underlying architecture of an SNS closely resembles the architecture of the NNS, but instead of being used to control the IC, an SNS is used to control a specific decentralized application that is running on the IC mainnet. Any dapp running on the mainnet can be handed over to an SNS, which results in the dapp being owned and controlled by a community. When an SNS takes control of a dapp, the dapp is then managed in the same way that the Internet Computer is managed by the NNS. When an SNS is launched, the ledger defines a unique token for the SNS that is then used by the community for participation in the SNS's governance. 

Reviewer: Maybe this is important in the tutorial, but why do we say "mainnet"? I wonder why it is important to call it mainnet, compared to just saying "ICP"?

Author: In the dev docs, we use "mainnet" to differentiate between local development and development that is deployed on the IC.

> The underlying architecture of an SNS closely resembles the architecture of the NNS, but instead of being used to control the IC, an SNS is used to control a specific decentralized application that is running on the IC mainnet. Any dapp running on the mainnet can be handed over to an SNS, which results in the dapp being owned and controlled by a community. When an SNS takes control of a dapp, the dapp is then managed in the same way that the Internet Computer is managed by the NNS. When an SNS is launched, the ledger defines a unique token for the SNS that is then used by the community for participation in the SNS's governance. 

Reviewer: ```suggestion
The underlying architecture of an SNS closely resembles the architecture of the NNS, but instead of being used to control the IC, an SNS is used to control a specific decentralized application that is running on the IC mainnet. Any dapp running on the mainnet can be handed over to an SNS, which results in the dapp being owned and controlled by a community. When an SNS takes control of a dapp, the dapp is then managed similarly to howthe Internet Computer is managed by the NNS. When an SNS is launched, a new unique token is created for the SNS that is then used by the community for participation in the SNS's governance. 
```

> The dapp's community can participate in the dapp's SNS through staking the SNS's native token in a neuron. Then, changes to the dapp are proposed to the community via an SNS proposal. SNS voting rights use the same format that voting on the NNS does; anyone can stake the SNS's native token into a neuron, and the neuron's voting power is proportional to the quantity of SNS tokens staked within the neuron, the length of their dissolve delay, and the amount of time since the neuron was created.

Reviewer: ```suggestion
Anyone, including the dapp's users, can participate in the dapp's SNS through staking the SNS's native token in an SNS neuron. Changes to the dapp are proposed to the community via an SNS proposal. SNS voting rights use the same format that voting on the NNS does; anyone can stake the SNS's native token into a neuron, and the neuron's voting power is proportional to the quantity of SNS tokens staked within the neuron, the length of their dissolve delay, and the amount of time since the neuron was created.
```
Wrt the last sentence: this is true, but every SNS can decide on the "weights" of the dissolve delay and age. Would this be worth mentioning?

Author: Resolving this since the SNS tutorial has been removed per my conversation with @LaraAS

## How an SNS works

> While the core architecture of an SNS resembles that of the NNS, the NNS contains additional canisters that provide important functionalities on the IC, such as the cycles minting canister that is responsible for creating cycles and the registry canister that contains the network's topology. In contrast, an SNS contains a few canisters that do not exist on the NNS, with the most notable canister being the decentralized swap canister used during the launch of the SNS. 

Reviewer: ```suggestion
While the core architecture of an SNS resembles that of the NNS, the NNS contains additional canisters that provide important functionalities on the IC, such as the cycles minting canister that is responsible for creating cycles and the registry canister that contains the network's topology. In contrast, an SNS contains a few canisters that do not exist on the NNS, with the most notable canister being the swap canister used for a decentralization swap during the launch of the SNS. 
```

> SNSs are provided as a system functionality. For an SNS to be launched, the NNS community must vote on and approve the SNS's original canister code. Once the SNS is launched, the code for the SNS's canisters is maintained by the IC and is changed through SNS proposals that are voted on by the community. All approved SNS canister versions are stored on an NNS canister known as the SNS Wasm Modules canister (SNS-W). Recall that all canisters are compiled into Wasm before being deployed on the IC.

Reviewer 2: This paragraph is a bit confusing to me since it seems to mix the SNS system canister and the dapp canisters that are controlled by a specific instance of a SNS.

Reviewer: I agree that this might be a bit confusing.. How about:
> SNSs are provided as a system functionality. 
All SNS canister code is approved by the NNS community and thus all SNSs follow the same upgrade path. 
Launching a new SNS is also approved by the NNS community, including the initial settings of the SNS and the conditions of the decentralization swap.

(Maybe the part about the Wasms and where they are stored is going into too much detail and can also be omited?)

Author: Resolving this since the SNS tutorial has been removed per my conversation with @LaraAS.

When an SNS is created, the SNS-W canister is responsible for deploying the latest version of the SNS's canisters. When an SNS is updated, an NNS proposal is used to add a new version of the SNS's canisters to the SNS-W canister. Then, the SNS's community decides via SNS proposal if they'd like to adopt the new, approved version into the SNS. Once approved via proposal, the updates are applied to the SNS's canisters. 

All SNSs are hosted on the SNS subnet, which is exclusively used to host SNSs. This allows for simplified verification for end users, since end users can verify that an SNS is running on the SNS subnet, and therefore trust that the SNS's underlying code has been approved by the NNS community. 

An SNS is comprised of the following canisters:

> - **SNS governance canister**: The governance canister defines who can participate in the SNS's governance decisions and automatically triggers the execution of all decisions. This canister stores the SNS proposals and a record of the neurons that define who the governance participants are. When a proposal is adopted, the governance system automatically triggers the execution of the proposal by calling a defined method. In most instances, the decision of the proposal is executed fully on-chain. 

Reviewer: ```suggestion
- **SNS governance canister**: This canister stores the SNS proposals and a record of the neurons that define who the governance participants are. When a proposal is adopted, the governance system automatically triggers the execution of the proposal by calling a defined method. In most instances, the decision of the proposal is executed fully on-chain. 
```

First sentence felt like a repetition

- **SNS ledger and archive canisters**: The ledger canister implements the ICRC-1 token standard to create a unique token for each SNS. To learn more about ICRC-1 tokens, refer to the previous module [4.2: ICRC-1 tokens](../level-4/4.2-icrc-tokens.md) where you deployed an ICRC-1 ledger then created an ICRC-1 token. This canister stores the SNS's ledger, which keeps track of how many accounts own SNS tokens, the amount owned by each account, and the history of transactions for the SNS's token. Since the canister has limited memory, to keep the ledger's full history, the ledger canister spawns archive canisters to store the ledger's block history. 

> - **SNS index canister**: The index canister keeps a record of which transactions are relevant for a given account. This provides the ability for wallets and other frontends to show all transactions for an account without requiring every frontend to implement this functionality themselves. When the NNS creates a new SNS, it will deploy the SNS's index canister as part of the deployment process. To deploy the index canister locally, follow the instructions found [here](/docs/current/developer-docs/integrations/icrc-1/icrc1-index-setup). 

Reviewer: ```suggestion
- **SNS index canister**: The index canister keeps a record of which transactions are relevant for a given account on the SNS ledger. This provides the ability for wallets and other frontends to show all transactions for an account without requiring every frontend to implement this functionality themselves.
```

I think the part wrt deployment holds for all these canisters, so I am not sure we have to make this explicit only for the index canister.

- **SNS root canister**: The root canister is responsible for upgrading the other SNS canisters and the dapp canisters that the SNS is responsible for governing. 

- **SNS decentralization swap canister**: The decentralization swap canister, also called the swap canister, is the primary canister involved when an SNS is launched. In a decentralization swap, users provide ICP tokens that are swapped for the SNS's native token. For an SNS to be launched, the decentralization swap must collect a certain amount of initial funding. This canister collects that initial funding and facilitates the distribution of neurons (and therefore the voting power) to the participants of the decentralization swap, making the SNS's governance decentralized.

> ### SNS asset canisters

Reviewer 2: Do we need this section here?

Author: Resolving this since the SNS tutorial has been removed per my conversation with @LaraAS

- #### Step 7: The SNS-W initializes the SNS canisters according to the initial parameters defined in step 1. 

- #### Step 8: The SNS decentralized swap begins. 

- #### Step 9: The SNS decentralized swap ends. 

- #### Step 10: The SNS decentralized swap finalizes. 

Detailed descriptions of each step can be found in the SNS documentation [here](/docs/current/developer-docs/integrations/sns/launching/launch-summary-1proposal). This tutorial will walk through these steps below. 

### SNS launch checklist

Before launching an SNS, the following checklist can be used to assure that your SNS is ready for launch:

- [x] Documentation:

    - [x] Tokenomics specification:

            - [x] Token utility: Define concise use-cases for the DAO's token.

            - [x] Initial token allocation amounts, including which groups or accounts receive how many tokens.  

            - [x] Voting power: Provide information about how voting power will be distributed at the SNS's genesis.

            - [x] Funding target: Maximum and minimum targets needed for the decentralized swap, including information about planned usage of raised funds. 

>             - [x] Use the [SNS tokenomics tool](https://docs.google.com/spreadsheets/u/0/d/1eSxkJl94jPt63CdOXH6ROy-WSkacW6P4qcAKMLrfBPc/edit) to analyze and document the SNS's proposed tokenomics. 

Reviewer 2: @bjoernek Do we still want to advertise the SNS tokenomics tool?

Reviewer 3: I think we should (but we should refer to the new tool on Github).

Author: Resolving this since the SNS tutorial has been removed per my conversation with @LaraAS

    - [x] Create a reproducible build: From the open source code, a reproducible build should be created. You can read more about reproducible builds [here](/docs/current/developer-docs/backend/reproducible-builds). 

    - [x] Test dapp operations with an SNS testflight: Before requesting an official SNS production launch, it is highly encouraged to test their deployed dapp's operations via SNS proposals using the SNS testflight suite. 

    - [x] Integrate an SNS frontend into the dapp: Decide whether the SNS's functionality will be integrated into the frontend of the dapp or not. One example of integration into the frontend is to include the SNS proposal voting on the dapp's frontend. 

    - [x] Test the SNS launch locally.

- [x] Community consultation:

    - [x] Publish the tokenomics, whitepaper, roadmap, and technical architecture documents.

>     - [x] Execute a community discussion on the Internet Computer forum to discuss the dapp's decentralization plans.

Reviewer 2: ```suggestion
    - [x] Start a community discussion on the Internet Computer forum presenting the documents at least 2 weeks before you plan to submit the proposal.
```
